cpp/src/lists/contains.cu

```diff
@@ -0,0 +1,155 @@
+/*
+ * Copyright (c) 2020, NVIDIA CORPORATION.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#include <thrust/logical.h>
+#include <cudf/column/column_factories.hpp>
+#include <cudf/detail/valid_if.cuh>
+#include <cudf/lists/contains.hpp>
+#include <cudf/lists/list_device_view.cuh>
+#include <cudf/lists/lists_column_device_view.cuh>
+#include <cudf/lists/lists_column_view.hpp>
+#include <cudf/scalar/scalar.hpp>
+#include <cudf/scalar/scalar_device_view.cuh>
+#include <cudf/utilities/type_dispatcher.hpp>
+#include <rmm/exec_policy.hpp>
+#include <type_traits>
+
+namespace cudf {
+namespace lists {
+
+namespace {
+
+auto CUDA_HOST_DEVICE_CALLABLE counting_iter(size_type n)
+{
+  return thrust::make_counting_iterator(n);
+}
+
+std::pair<rmm::device_buffer, size_type> construct_null_mask(
+  cudf::detail::lists_column_device_view const& d_lists,
+  cudf::scalar const& skey,
+  bool input_column_has_nulls,
+  rmm::cuda_stream_view stream,
+  rmm::mr::device_memory_resource* mr)
+{
+  using namespace cudf;
+  using namespace cudf::detail;
+
+  if (!skey.is_valid(stream)) {
+    return std::make_pair(cudf::create_null_mask(d_lists.size(), mask_state::ALL_NULL, mr),
+                          d_lists.size());
+  }
+
+  return cudf::detail::valid_if(
+    counting_iter(0), counting_iter(d_lists.size()), [d_lists] __device__(auto const& row_index) {
+      auto list = cudf::list_device_view(d_lists, row_index);
+      if (list.is_null()) { return false; }
```

Comment:
@davidwendt, et al, what is the prescribed way to switch this to use `list.is_null_no_check()`, so that the existence of the null mask is checked only once?

Consider that there will be another `construct_null_mask()` overload where `skey` is a column. There will then be *four* combinations to consider, which will likely obfuscate the code.